process/src/controllers/controller.serialport.ts

```diff
@@ -1,214 +1,305 @@
+      this._reader.removeAllListeners();
+      this._port.removeAllListeners();
+      this._port.unpipe(this._reader);
+      this._port.close((error: Error | null | undefined) => {
+        this._logger.verbose("Successfully closed!");
+        this._port = undefined;
+        this._reader = undefined;
+        if (error) {
+          return reject(error);
+        }
+        resolve();
+      });
+    });
+  }
 
-    public open(): Promise<void> {
-        return new Promise((resolve, reject) => {
-            if (this._port !== undefined) {
-                return reject(new Error(this._logger.error(`Port is already exist.`)));
-            }
-            const optErrs: Error | undefined = this._getOptionsErrors(this._options);
-            if (optErrs) {
-                return reject(new Error(this._logger.error(`Error opening port, because options aren't valid: ${optErrs.message}`)));
-            }
-            this._port = new SerialPort(this._options.path, this._options.options, (error: Error | undefined | null) => {
-                if (error) {
-                    return reject(new Error(this._logger.error(`Fail to open serial port due error: ${error.message}`)));
-                }
-                this._refPortEvents(this._port as SerialPort);
-                this._logger.error(`Connection to port "${this._options.path}" is successfull`);
-                resolve();
-            });
-        });
-    }
+  public open(): Promise<void> {
+    return new Promise((resolve, reject) => {
+      if (this._port !== undefined) {
+        return reject(new Error(this._logger.error(`Port is already exist.`)));
+      }
+      const optErrs: Error | undefined = this._getOptionsErrors(this._options);
+      if (optErrs) {
+        return reject(
+          new Error(
+            this._logger.error(
+              `Error opening port, because options aren't valid: ${optErrs.message}`
+            )
+          )
+        );
+      }
+      this._port = new SerialPort(
+        this._options.path,
+        this._options.options,
+        (error: Error | undefined | null) => {
+          if (error) {
+            return reject(
+              new Error(
+                this._logger.error(
+                  `Fail to open serial port due error: ${error.message}`
+                )
+              )
+            );
+          }
+          this._refPortEvents(this._port as SerialPort);
+          this._logger.error(
+            `Connection to port "${this._options.path}" is successfull`
+          );
+          resolve();
+        }
+      );
+    });
+  }
 
-    public write(chunk: Buffer | string): Promise<void> {
-        return new Promise((resolve, reject) => {
-            if (this._port === undefined) {
-                return reject(new Error(this._logger.error(`Fail to write in port ${this._options.path} because port isn't inited`)));
+  public write(chunk: Buffer | string): Promise<void> {
+    return new Promise((resolve, reject) => {
```

Comment:
This method still has a problems with resolving and rejecting. 
Resolve now is in timer callback; but rejection is located in nested `this._port.write` callback

Reply:
please double-check `all` promises in the solution:
- root promise always should be resolved/rejected
- we should reject or resolve it only once
- if we want to make root promise to be only resolved (like in destroying methods) - use `finally` instead `then`